runtime/lua/vim/lsp.lua

```diff
@@ -811,78 +811,119 @@ local function once(fn)
-      string.format("on_lines bufnr: %s, changedtick: %s, firstline: %s, lastline: %s, new_lastline: %s, old_byte_size: %s, old_utf32_size: %s, old_utf16_size: %s",
-      bufnr, changedtick, firstline, lastline, new_lastline, old_byte_size, old_utf32_size, old_utf16_size),
-      nvim_buf_get_lines(bufnr, firstline, new_lastline, true)
+      string.format("on_bytes bufnr: %s, changedtick: %s, start_row: %s, " .. 
+                    "start_column: %s, byte_offset: %s, old_end_row: %s, " ..
+                    "old_end_column: %s, old_end_byte_length: %s, " ..
+                    "new_end_row: %s, new_end_column: %s, new_end_byte_length: %s",
+                    bufnr, changedtick, start_row, start_column, byte_offset,
+                    old_end_row, old_end_column, old_end_byte_length, new_end_row, 
+                    new_end_column, new_end_byte_length),
+      buf_range_to_text(bufnr, {
+          start_row = start_row,
+          start_column = start_column,
+          old_end_row=old_end_row,
+          old_end_column = old_end_column,
+          })
     )
 
     -- Don't do anything if there are no clients attached.
     if tbl_isempty(all_buffer_active_clients[bufnr] or {}) then
       return
     end
 
-    util.buf_versions[bufnr] = changedtick
-    -- Lazy initialize these because clients may not even need them.
     local incremental_changes = once(function(client)
-      local size_index = encoding_index[client.offset_encoding]
-      local length = select(size_index, old_byte_size, old_utf16_size, old_utf32_size)
-      local lines = nvim_buf_get_lines(bufnr, firstline, new_lastline, true)
+      local text = buf_range_to_text(bufnr, {
+                      start_row = start_row, 
+                      start_column = start_column,
+                      old_end_row=old_end_row,
+                      old_end_column = old_end_column,
+                      })
+
+      vim.notify(vim.inspect({
+        range = {
+          -- TODO: UTF??
+          -- TODO: Is it old end or new end?
+          start = { line = start_row, character = start_column - 1 };
+          ["end"] = { line = start_row+old_end_row, character = start_column+old_end_column - 1};
+        };
+        text = text;
+      }))
 
       -- This is necessary because we are specifying the full line including the
       -- newline in range. Therefore, we must replace the newline as well.
-      if #lines > 0 then
-       table.insert(lines, '')
-      end
+      -- if #lines > 0 then
+      --  table.insert(lines, '')
+      -- end
       return {
         range = {
-          start = { line = firstline, character = 0 };
-          ["end"] = { line = lastline, character = 0 };
+          -- TODO: UTF??
+          -- TODO: Is it old end or new end?
+          start = { line = start_row, character = start_column - 1 };
+          ["end"] = { line = start_row+old_end_row, character = start_column+old_end_column - 1};
         };
-        rangeLength = length;
```

Comment:
isn't `rangeLength` still needed?

Reply:
That was recently deprecated https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_didChange

Reply:
```
	/**
	 * The optional length of the range that got replaced.
	 *
	 * @deprecated use range instead.
	 */
	rangeLength?: uinteger;
```

Reply:
And servers (the common ones) have been updated to not require it?

Reply:
I don't think sumneko, RLS, HLS, or vscode-language server node (pyright, tsserver, yamlls, etc) require it based on a cursory look at the sources, but I can add it for backwards compatibility

Reply:
Well in that case we can wait and see if someone complains :)